lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -1150,6 +1189,128 @@ void SESERegionBuilder::ensureSingleExitFromLoops() {
   }
 }
 
+void SingleExitLoopTransformer::unrollLoopBody() {
+  BasicBlockCloner cloner(*currentFn);
+  // Setup cloner so that newHeader's argument's are replaced with values in
+  // preheader.
+  SILBasicBlock *newHeader = loop->getHeader();
+  auto preheaderTermInst = dyn_cast<BranchInst>(preheader->getTerminator());
+  assert(preheaderTermInst && "Preheader of a loop has a non-branch terminator");
+  for (unsigned argIndex = 0; argIndex < oldHeaderNumArgs; ++argIndex) {
+    auto preHeaderArg = preheaderTermInst->getArg(argIndex);
+    auto newHeaderArg = newHeader->getArgument(argIndex);
+    cloner.updateValueMap(newHeaderArg, preHeaderArg);
+  }
+  // Clone everything except the new header. We should traverse the
+  // blocks in depth first order to ensure values are cloned before they are used.
+  SmallPtrSet<SILBasicBlock *, 32> worklist;
+  SmallVector<SILBasicBlock *, 32> initializedBlocks;
+  worklist.insert(header);
+  while (!worklist.empty()) {
+    SILBasicBlock *current = *worklist.begin();
+    worklist.erase(current);
+    cloner.initBlock(current);
+    initializedBlocks.push_back(current);
+    for (SILBasicBlock *succ : current->getSuccessorBlocks()) {
+      // Skip if succ is not a part of the loop, is already cloned, or
+      // is the new preheader.
+      if (!loop->contains(succ) || cloner.remapBasicBlock(succ) != succ ||
+          succ == newHeader) {
+        continue;
+      }
+      worklist.insert(succ);
+    }
+  }
+  for (SILBasicBlock *bb : initializedBlocks) {
+    cloner.cloneBlock(bb);
+  }
+
+  // Get the clone for the original and new header.
+  SILBasicBlock *clonedHeader = cloner.remapBasicBlock(header);
```

Comment:
> Get the clone for the original and new header.

This isn't always a clone, right? The implementation of `BasicBlockCloner::remapBasicBlock` returns the original argument if it's not in the dictionary.